src/hu_coordinates.c

```diff
@@ -0,0 +1,221 @@
+//
+// Copyright(C) 2022 Ryan Krafnick
+// Copyright(C) 2024 ceski
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//      Advanced Coordinates HUD Widget
+//
+
+#include <math.h>
+
+#include "hu_coordinates.h"
+#include "m_misc.h"
+#include "p_mobj.h"
+#include "v_video.h"
+
+#define THRESH_M1 15.11
+#define THRESH_M2 19.35
+#define THRESH_M3 21.37
+
+#define THRESH_D1 16.67
+#define THRESH_D2 21.35
+#define THRESH_D3 23.58
+
+#define BLOCK_COLOR(x, a, b, c) \
+    (x >= c ? CR_RED : x >= b ? CR_BLUE1 : x >= a ? CR_GREEN : CR_GRAY)
+
+#define WIDGET_WIDTH 18
+
+typedef struct
+{
+    boolean negative;
+    int base;
+    int frac;
+} split_fixed_t;
+
+typedef struct
+{
+    int base;
+    int frac;
+} split_angle_t;
+
+static double CalcMomentum(const mobj_t *mo)
+{
+    const double mx = FIXED2DOUBLE(mo->momx);
+    const double my = FIXED2DOUBLE(mo->momy);
+
+    return sqrt(mx * mx + my * my);
+}
+
+static double CalcDistance(const mobj_t *mo)
+{
+    const double dx = FIXED2DOUBLE(mo->x - mo->oldx);
+    const double dy = FIXED2DOUBLE(mo->y - mo->oldy);
+
+    return sqrt(dx * dx + dy * dy);
+}
+
+static split_fixed_t SplitFixed(fixed_t x)
+{
+    split_fixed_t result;
+
+    result.negative = x < 0;
+    result.base = x >> FRACBITS;
+    result.frac = x & 0xFFFF;
```

Comment:
Isn't this `FRACUNIT - 1`?

Reply:
Changed here: https://github.com/fabiangreffrath/woof/pull/1760/commits/1215df0d6754ac47256313cfe376b79890976998

Reply:
> Can we do something like `FRACMASK = FRACUNIT - 1`?

Oh, I didn't see this edit. How about we add a macro for that in a different PR?

Reply:
Sure, that's ok.